src/sentry/workflow_engine/handlers/detector/stateful.py

```diff
@@ -248,29 +261,35 @@ def get_state_data(
         return results
 
 
+DetectorThresholds = dict[DetectorPriorityLevel, int]
+
+
 class StatefulDetectorHandler(
     Generic[DataPacketType, DataPacketEvaluationType],
     DetectorHandler[DataPacketType, DataPacketEvaluationType],
     abc.ABC,
 ):
-    def __init__(self, detector: Detector):
+    """
+    Stateful Detectors are provided as a base class for new detectors that need to track state.
+    """
+
+    thresholds: DetectorThresholds = {}
+
+    def __init__(self, detector: Detector, thresholds: DetectorThresholds | None = None):
         super().__init__(detector)
-        # if detector.config.get("thresholds"):
-        #     raise Exception(
-        #         "Stateful detectors are required to have `thresholds` set in the config"
-        #     )
 
-        counter_names = self.counter_names or []
-        self.state_manager = DetectorStateManager(detector, counter_names)
+        self._thresholds: DetectorThresholds = {
+            **{level: 1 for level in DetectorPriorityLevel},  # Default to 0 for all levels
+            **(self.thresholds or {}),  # Allow each handler to override
+            **(thresholds or {}),  # Allow each instance to override
+        }
 
-    @property
-    @abc.abstractmethod
-    def counter_names(self) -> list[str]:
-        """
-        The names of the counters that this detector tracks. This is used to build the redis keys for
-        storing counter values.
-        """
-        pass
+        # Merge all kinds of counters together for the state manager
```

Comment:
Nit: These are already merged in `self._thresholds`

Reply:
The way i was thinking about this is that there might be additional counters that the implementer wants to track in their detector handler. this is meant to give the space to allow for that override fairly easily -- but yeah, i feel like this is probs an over optimization, i can simplify to just having threshold counters on the stateful detector for now.